flowcontrol/fixed.go

```diff
@@ -3,35 +3,38 @@ package flowcontrol
 import (
 	"context"
 	"fmt"
-	"math"
 
 	"golang.org/x/sync/semaphore"
 )
 
 // Returns a FlowLimiter that enforces a fixed limit on the total size of
 // outstanding messages.
 func NewFixedLimiter(size int64) FlowLimiter {
-	return (*fixedLimiter)(semaphore.NewWeighted(size))
+	return &fixedLimiter{
+		size: size,
+		sem:  semaphore.NewWeighted(size),
+	}
 }
 
-type fixedLimiter semaphore.Weighted
+type fixedLimiter struct {
+	size int64
+	sem  *semaphore.Weighted
+}
 
 func (fl *fixedLimiter) StartMessage(ctx context.Context, size uint64) (gotResponse func(), err error) {
-	if size > math.MaxInt64 {
-		// semaphore.Weighted expects an int64, so we need to check the bounds.
-		return nil, fmt.Errorf(
-			"StartMessage(): limit %v is too large (max %v)",
-			size, int64(math.MaxInt64),
-		)
+	// HACK:  avoid dead-locking if the size of the message exceeds the maximum
+	//        reservation on the semaphore. We can't return an error because it
+	//        is currently ignored by the caller.
+	if int64(size) > fl.size {
+		const s = "StartMessage(): message size %d is too large (max %d)"
+		panic(fmt.Sprintf(s, size, fl.size))
```

Comment:
IMO factoring the format string out into a constant like this just makes things more confusing. There's also probably a linter out there that will yell at you for this...

Reply:
Done